.github/workflows/build_and_test.yml

```diff
@@ -1066,6 +1066,70 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
+      - name: Create git tag and GitHub release
+        if: github.ref_name == 'release' || github.ref_name == 'release-proxy' || github.ref_name == 'release-compute'
+        uses: actions/github-script@v7
+        with:
+          retries: 5
+          script: |
+            const tag = "${{ needs.tag.outputs.build-tag }}";
+
+            try {
+              const existingRef = await github.rest.git.getRef({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                ref: `tags/${tag}`,
+              });
+
+              if (existingRef.data.object.sha !== context.sha) {
+                throw new Error(`Tag ${tag} already exists but points to a different commit (expected: ${context.sha}, actual: ${existingRef.data.object.sha}).`);
+              }
+
+              console.log(`Tag ${tag} already exists and points to ${context.sha} as expected.`);
+            } catch (error) {
+              if (error.status !== 404) {
+                throw error;
+              }
+
+              console.log(`Tag ${tag} does not exist. Creating it...`);
+              await github.rest.git.createRef({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                ref: `refs/tags/${tag}`,
+                sha: context.sha,
+              });
+              console.log(`Tag ${tag} created successfully.`);
```

Comment:
Optional: I'm a bit hesitant about having the main logic in `catch` block. 
Should we try to create a reference first, and if it return 409 (from [GitHub API docs](https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#create-a-reference--status-codes)), then to check if it's for the same tag and fail with an error if not?
The approach won't work in the same way for `createRelease`, unfortunately

Reply:
I have played around with different variants of this, and no extraction of logic out of the catch block improved readability in a substantial amount.